pincer/utils/event_mgr.py

```diff
@@ -0,0 +1,287 @@
+
+
+class _LoopMgr(_Processable):
+    """
+    Parameters
+    ----------
+    event_name : str
+        The name of the event.
+    check : Optional[Callable[[Any], bool]]
+        ``can_be_set`` only returns true if this function returns true.
+        Will be ignored if set to None.
+
+    Attributes
+    ----------
+    can_expand : bool
+        Whether the queue is allowed to grow. Turned to false once the
+        EventMgr's timer runs out.
+    events : :class:`collections.deque`
+        Qeue of events to be processed.
+    wait : :class:`asyncio.Event`
+        Used to make ``get_next()` wait for the next event.
+    """
+
+    def __init__(self, event_name: str, check: CheckFunction) -> None:
+        self.event_name = event_name
+        self.check = check
+
+        self.can_expand = True
+        self.events = deque()
+        self.wait = Event()
+
+    def process(self, event_name: str, *args):
+        if not self.can_expand:
+            return
+
+        if self.matches_event(event_name, *args):
+            self.events.append(args)
+            self.wait.set()
+
+    async def get_next(self):
+        """
+        Returns the next item if the queue. If there are no items in the queue,
+        it will return the next event that happens.
+        """
+        if len(self.events) == 0:
+            if not self.can_expand:
+                raise _LoopEmptyError
+
+            self.wait.clear()
+            await self.wait.wait()
+            return self.events.popleft()
+        else:
+            return self.events.popleft()
+
+
+class EventMgr:
+    """
+    Attributes
+    ----------
+    event_list : List[_DiscordEvent]
+        The List of events that need to be processed.
+    """
+
+    def __init__(self):
+        self.event_list: _Processable = []
+
+    def process_events(self, event_name, *args):
+        """
+        Parameters
+        ----------
+        event_name : str
+            The name of the event to be processed.
+        *args : Any
+            The arguments returned from the middleware for this event.
+        """
+        for event in self.event_list:
+            event.process(event_name, *args)
+
+    async def wait_for(
+        self,
+        event_name: str,
+        check: CheckFunction,
+        timeout: Union[float, None]
+    ) -> Any:
+        """
+        Parameters
+        ----------
+        event_name : str
+            The type of event. It should start with `on_`. This is the same
+            name that is used for @Client.event.
+        check : Union[Callable[[Any], bool], None]
+            This function only returns a value if this return true.
+        timeout: Union[float, None]
+            Amount of seconds before timeout. Use None for no timeout.
+
+        Returns
+        ------
+        Any
+            What the Discord API returns for this event.
+        """
+
+        event = _Event(
+            event_name=event_name,
+            check=check
+        )
```

Comment:
Don't think it's the best idea to use keyword arguments if they're not required here
```suggestion
        event = _Event(event_name, check)
```